// @trezor/protocol/lib/protocol-thp/messages/protobufDefinitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProtobufDefinitions = void 0;
const getProtobufDefinitions = () => ({
    ThpPairingMethod: { values: { SkipPairing: 1, CodeEntry: 2, QrCode: 3, NFC: 4 } },
    ThpDeviceProperties: {
        fields: {
            internal_model: { type: 'string', id: 1 },
            model_variant: { type: 'uint32', id: 2 },
            protocol_version_major: { type: 'uint32', id: 3 },
            protocol_version_minor: { type: 'uint32', id: 4 },
            pairing_methods: {
                rule: 'repeated',
                type: 'ThpPairingMethod',
                id: 5,
                options: { packed: false },
            },
        },
    },
    ThpHandshakeCompletionReqNoisePayload: {
        fields: { host_pairing_credential: { type: 'bytes', id: 1 } },
    },
    ThpCreateNewSession: {
        fields: {
            passphrase: { type: 'string', id: 1 },
            on_device: { type: 'bool', id: 2 },
            derive_cardano: { type: 'bool', id: 3 },
        },
    },
    ThpPairingRequest: { fields: { host_name: { type: 'string', id: 1 } } },
    ThpPairingRequestApproved: { fields: {} },
    ThpSelectMethod: { fields: { selected_pairing_method: { type: 'ThpPairingMethod', id: 1 } } },
    ThpPairingPreparationsFinished: { fields: {} },
    ThpCodeEntryCommitment: { fields: { commitment: { type: 'bytes', id: 1 } } },
    ThpCodeEntryChallenge: { fields: { challenge: { type: 'bytes', id: 1 } } },
    ThpCodeEntryCpaceTrezor: { fields: { cpace_trezor_public_key: { type: 'bytes', id: 1 } } },
    ThpCodeEntryCpaceHostTag: {
        fields: { cpace_host_public_key: { type: 'bytes', id: 1 }, tag: { type: 'bytes', id: 2 } },
    },
    ThpCodeEntrySecret: { fields: { secret: { type: 'bytes', id: 1 } } },
    ThpQrCodeTag: { fields: { tag: { type: 'bytes', id: 1 } } },
    ThpQrCodeSecret: { fields: { secret: { type: 'bytes', id: 1 } } },
    ThpNfcTagHost: { fields: { tag: { type: 'bytes', id: 1 } } },
    ThpNfcTagTrezor: { fields: { tag: { type: 'bytes', id: 1 } } },
    ThpCredentialRequest: {
        fields: {
            host_static_pubkey: { type: 'bytes', id: 1 },
            autoconnect: { type: 'bool', id: 2 },
            credential: { type: 'bytes', id: 3 },
        },
    },
    ThpCredentialResponse: {
        fields: {
            trezor_static_pubkey: { type: 'bytes', id: 1 },
            credential: { type: 'bytes', id: 2 },
        },
    },
    ThpEndRequest: { fields: {} },
    ThpEndResponse: { fields: {} },
    MessageType: {
        values: {
            ThpCreateNewSession: 1000,
            ThpPairingRequest: 1006,
            ThpPairingRequestApproved: 1007,
            ThpSelectMethod: 1008,
            ThpPairingPreparationsFinished: 1009,
            ThpCredentialRequest: 1010,
            ThpCredentialResponse: 1011,
            ThpEndRequest: 1012,
            ThpEndResponse: 1013,
            ThpCodeEntryCommitment: 1016,
            ThpCodeEntryChallenge: 1017,
            ThpCodeEntryCpaceTrezor: 1018,
            ThpCodeEntryCpaceHostTag: 1019,
            ThpCodeEntrySecret: 1020,
            ThpQrCodeTag: 1024,
            ThpQrCodeSecret: 1025,
            ThpNfcTagHost: 1032,
            ThpNfcTagTrezor: 1033,
        },
    },
});
exports.getProtobufDefinitions = getProtobufDefinitions;
